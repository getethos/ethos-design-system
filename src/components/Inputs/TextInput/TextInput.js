import React from 'react'
import PropTypes from 'prop-types'
import useRequired from '../../../hooks/useRequired.js'
import useErrorMessage from '../../../hooks/useErrorMessage.js'
import useInvalid from '../../../hooks/useInvalid.js'

import { Body, COLORS } from '../../index'

// import './TextInput.scss'

/* @getethos/design-system/TextInput.js

/**
 * WIP
 *
 * @param  {String}   props.name        Input name and htmlFor prop for label
 * @param  {String}   props.labelCopy   User-visible text of label for input
 * @param  {Function} props.validator   Function for validating input
 * @param  {Boolean}  props.disabled  
 */

function PrivateTextInput({ disabled, name, labelCopy, validator, ...rest }) {
  // Verify that all required props were supplied
  const [includesRequired] = useRequired(['data-tid', 'name', 'labelCopy'])
  let allRelevantProps = Object.assign({}, rest, {
    name: name,
    labelCopy: labelCopy,
  })
  includesRequired(allRelevantProps)

  // Verify that no invalid props were supplied
  const [includesInvalid] = useInvalid(
    Object.keys(PrivateTextInput.PUBLIC_PROPS)
  )
  includesInvalid(rest)

  // Set up validation hooks
  const [getError, setError, validate] = useErrorMessage(validator)

  // Generate list of css classes
  const classNames = ['TextInput']

  const onChange = (syntheticReactEvent) => {
    const errMsg = validate(syntheticReactEvent.target.value)
    if (errMsg.length) {
      setError(errMsg)
    } else {
      setError('')
    }
  }

  return (
    <>
      <Body.Regular400
        element="label"
        htmlFor={name}
        color={COLORS.GRAY_PRIMARY}
      >
        {labelCopy}
      </Body.Regular400>
      <input
        type="text"
        className={classNames.join(' ')}
        disabled={disabled}
        name={name}
        onChange={onChange}
        data-tid={rest['data-tid']}
      />
      {getError()}
    </>
  )
}

PrivateTextInput.PUBLIC_PROPS = {
  'data-tid': PropTypes.string.isRequired,
  disabled: PropTypes.bool,
  name: PropTypes.string.isRequired,
  labelCopy: PropTypes.string.isRequired,
  validator: PropTypes.func,
}

PrivateTextInput.propTypes = {
  ...PrivateTextInput.PUBLIC_PROPS,
}

function TextInputFactory(privateProps) {
  const PublicTextInputComponent = (downstreamProps) => {
    return <PrivateTextInput {...downstreamProps} {...privateProps} />
  }

  return PublicTextInputComponent
}

export const TextInput = TextInputFactory()
