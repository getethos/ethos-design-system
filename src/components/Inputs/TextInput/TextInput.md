```jsx
import validateMinMaxFactory from '../../../validators/validateMinMax'
;<TextInput
  name="example"
  labelCopy="Validation happens after first blur ('touched')â€”Value's length % 2"
  data-tid="the-text-input"
  validator={(x) => {
    const minMaxErr = validateMinMaxFactory(5, 20)(x)
    if (!!minMaxErr) return minMaxErr
    return x.length % 2 ? 'Text does not have an even number of characters' : ''
  }}
/>
```
