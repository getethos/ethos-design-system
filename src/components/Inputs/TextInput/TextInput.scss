.TextInput {
  // Same as Type.scss Body. Should we use a mixin?
  $font-size: 19;
  $line-height: 24;
  font-size: $font-size * 1px;
  line-height: $line-height / $font-size;
  width: 100%;
  font-family: var(--Theinhardt-font-stack);
  box-sizing: border-box;
  background-color: var(--White);
  border-style: solid;
  border-width: 1px;
  border-color: var(--GrayStrokeAndDisabled--translucent);
  // TODO: Is there a better way to achieve this?
  padding: 17px 16px 13px;
  color: var(--GrayPrimary--opaque);
  border-radius: 0;
  transition: border-color 250ms ease;
  outline: 0;
  margin: 0;
}

.TextInput:hover {
  background-color: var(--GrayLightHover--translucent);
}

.TextInput:focus {
  border-color: var(--GrayPrimary--opaque);
  background-color: var(--White);
}

.TextInput::placeholder {
  // It's $gray-64 / #a3a3a3 in ethos, but going to try to limit # of grays
  color: var(--GrayStrokeAndDisabled--opaque);
}

.TextInput:disabled {
  color: var(--White);
  background-color: var(--GrayStrokeAndDisabled--translucent);
  border-color: transparent;
}
