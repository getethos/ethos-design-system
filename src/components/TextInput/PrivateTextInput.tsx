import * as PropTypes from 'prop-types'
import * as React from 'react'
import { useEffect, useState } from 'react'
import restrict from '../../helpers/restrict.js'
import useErrorMessage from '../../hooks/useErrorMessage'
import useInputValidation from '../../hooks/useInputValidation'
import useInvalid from '../../hooks/useInvalid.js'
import useRequired from '../../hooks/useRequired.js'
import errorStyles from '../Errors.module.scss'
import { PublicFormProps } from '../Form/Form'
import { InputLabel } from '../InputLabel'
import styles from './TextInput.module.scss'
const PUBLIC_PROPS = {
  type: PropTypes.string,
  'data-tid': PropTypes.string.isRequired,
  disabled: PropTypes.bool,
  name: PropTypes.string.isRequired,
  allCaps: PropTypes.bool,
  /** text transform capitalize label */
  capitalize: PropTypes.bool,
  initialValue: PropTypes.string,
  labelCopy: PropTypes.string,
  validator: PropTypes.func,
  placeholder: PropTypes.string,
  onBlur: PropTypes.func,
  onFocus: PropTypes.func,
  restrictIllegal: PropTypes.bool,
}
type PrivateTextInputProps = PublicFormProps & {
  type?: string
  disabled?: boolean
  allCaps?: boolean
  capitalize?: boolean
  initialValue?: string
  currentValue?: string
  currentError?: any
  labelCopy?: string
  placeholder?: string
  onBlur?: (...args: any[]) => any
  onFocus?: (...args: any[]) => any
  restrictIllegal?: boolean
}
export const PrivateTextInput: React.FC<PrivateTextInputProps> & {
  PUBLIC_PROPS: typeof PUBLIC_PROPS
} = ({
  type,
  disabled,
  name,
  labelCopy,
  allCaps,
  capitalize,
  formChangeHandler,
  validator,
  initialValue,
  currentValue,
  currentError,
  setFieldTouched,
  restrictIllegal,
  ...rest
}) => {
  // Verify that all required props were supplied
  const [includesRequired] = useRequired(['data-tid', 'name'])
  let allRelevantProps = Object.assign({}, rest, {
    name: name,
    allCaps: allCaps,
    capitalize: capitalize,
  })
  includesRequired(allRelevantProps)
  // Verify that no invalid props were supplied
  const [includesInvalid] = useInvalid(
    Object.keys(PrivateTextInput.PUBLIC_PROPS)
  )
  includesInvalid(rest)
  // Set up validation hooks
  const [getError, setError, , validate] = useErrorMessage(validator)
  const [value, setValue] = useState(currentValue || initialValue || '')
  const [touched, setTouched] = useState(initialValue ? true : false)
  const [doValidation] = useInputValidation({
    validate,
    setError,
    formChangeHandler,
  })
  const onChange = (ev) => {
    const val = ev.target.value
    const restrictedVal = restrictIllegal ? restrict(val) : val
    setValue(restrictedVal)
    // We call setTouched in onBlur, so can reliably call getter here
    doValidation(restrictedVal, touched)
  }
  const setAllTouched = () => {
    // We set touched to change the react state, but it's async and
    // processing still, so, we use a flag for doValidation
    setTouched(true)
    // Also tell the form we've been touched
    if (setFieldTouched) {
      setFieldTouched(true)
    }
  }
  // Initial Value aka prefilledâ€”are considered "touched", but must prevalidate
  // which will in turn update the internal form state as to their validity
  useEffect(() => {
    if (!!formChangeHandler && initialValue) {
      doValidation(initialValue, true)
    }
  }, [])
  const onBlur = (ev) => {
    setAllTouched()
    doValidation(ev.target.value, true)
  }
  const getClasses = () => {
    return getError(currentError, touched)
      ? `${styles.TextInput} ${errorStyles.Error}`
      : `${styles.TextInput}`
  }
  return (
    <>
      {labelCopy && (
        <InputLabel
          name={name}
          labelCopy={labelCopy}
          allCaps={allCaps}
          capitalize={capitalize}
        />
      )}
      <input
        type={type}
        className={getClasses()}
        disabled={disabled}
        name={name}
        placeholder={rest.placeholder}
        onChange={onChange}
        onBlur={onBlur}
        value={value}
        data-tid={rest['data-tid']}
        aria-label={name} // https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-label_attribute
      />
      {getError(currentError, touched)}
    </>
  )
}
PrivateTextInput.PUBLIC_PROPS = PUBLIC_PROPS
PrivateTextInput.defaultProps = {
  type: 'text',
  placeholder: '',
  restrictIllegal: true,
}
