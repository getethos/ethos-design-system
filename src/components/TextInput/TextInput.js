import React, { useState, useEffect } from 'react'
import PropTypes from 'prop-types'

import { InputLabel } from '../InputLabel'
import useRequired from '../../hooks/useRequired.js'
import useErrorMessage from '../../hooks/useErrorMessage.js'
import useInvalid from '../../hooks/useInvalid.js'
import useInputValidation from '../../hooks/useInputValidation.js'
import restrict from '../../helpers/restrict.js'
import styles from './TextInput.module.scss'
import errorStyles from '../Errors.module.scss'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { VALID_ICONS } from '../../helpers/constants'

/**
 * @param type
 * @param  {string}   name        Input name and htmlFor prop for label
 * @param  {string}   labelCopy   User-visible text of label for input
 * @param  {boolean}  allCaps     Whether to text-trasform: uppercase
 * @param  {Function} validator   Function for validating input
 * @param  {boolean}  disabled
 * @param capitalize
 * @param formChangeHandler
 * @param initialValue
 * @param currentValue
 * @param currentError
 * @param setFieldTouched
 * @param restrictIllegal
 * @param  {string}   autoComplete  Autocomplete label
 * @param {string} [classOverrides]
 * @param {string} [labelColor]
 * @param {string} [labelWeight]
 * @param {string} [labelClasses]
 * @param rest
 */

function PrivateTextInput({
  type,
  disabled,
  name,
  labelCopy,
  allCaps,
  capitalize,
  formChangeHandler,
  validator,
  initialValue,
  currentValue,
  currentError,
  setFieldTouched,
  restrictIllegal,
  autoComplete,
  classOverrides,
  labelColor,
  labelWeight,
  labelClasses,
  icon,
  ...rest
}) {
  // Verify that all required props were supplied
  const [includesRequired] = useRequired(['data-tid', 'name'])
  let allRelevantProps = Object.assign({}, rest, {
    name: name,
    allCaps: allCaps,
    capitalize: capitalize,
  })
  includesRequired(allRelevantProps)

  // Verify that no invalid props were supplied
  const [includesInvalid] = useInvalid(
    Object.keys(PrivateTextInput.PUBLIC_PROPS)
  )
  includesInvalid(rest)

  // Set up validation hooks
  const [getError, setError, , validate] = useErrorMessage(validator)

  const [value, setValue] = useState(currentValue || initialValue || '')

  const [touched, setTouched] = useState(initialValue ? true : false)

  const [doValidation] = useInputValidation({
    validate,
    setError,
    formChangeHandler,
  })

  const onChange = (ev) => {
    const val = ev.target.value
    const restrictedVal = restrictIllegal ? restrict(val) : val
    setValue(restrictedVal)

    // We call setTouched in onBlur, so can reliably call getter here
    doValidation(restrictedVal, touched)
  }

  const setAllTouched = () => {
    // We set touched to change the react state, but it's async and
    // processing still, so, we use a flag for doValidation
    setTouched(true)
    // Also tell the form we've been touched
    if (setFieldTouched) {
      setFieldTouched(true)
    }
  }

  // Initial Value aka prefilledâ€”are considered "touched", but must prevalidate
  // which will in turn update the internal form state as to their validity
  useEffect(() => {
    if (!!formChangeHandler && initialValue) {
      doValidation(initialValue, true)
    }
  }, [])

  const onBlur = (ev) => {
    setAllTouched()
    doValidation(ev.target.value, true)
  }

  const getClasses = () => {
    let returnClasses

    if (getError(currentError, touched)) {
      returnClasses = `${styles.TextInputCommon} ${errorStyles.Error}`
    } else if (classOverrides) {
      returnClasses = `${styles.TextInputCommon} ${classOverrides}`
    } else {
      returnClasses = `${styles.TextInputCommon} ${styles.TextInputStylable}`
    }
    // hasIcon class indicates the text input has icon, to give more padding-right in stylings,to prevent overlapping between icon and long input
    if (Object.keys(VALID_ICONS).includes(icon)) {
      returnClasses += ` ${styles.hasIcon}`
    }
    return returnClasses
  }

  return (
    <>
      {labelCopy && (
        <InputLabel
          name={name}
          labelCopy={labelCopy}
          labelColor={labelColor}
          labelWeight={labelWeight}
          labelClasses={labelClasses}
          allCaps={allCaps}
          capitalize={capitalize}
        />
      )}
      <div className={styles.TextInputWrapper}>
        <input
          type={type}
          className={getClasses()}
          disabled={disabled}
          name={name}
          placeholder={rest.placeholder}
          onChange={onChange}
          onBlur={onBlur}
          value={value}
          maxLength={rest.maxLength}
          data-tid={rest['data-tid']}
          aria-label={name} // https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-label_attribute
          autoComplete={autoComplete}
        />
        {Object.keys(VALID_ICONS).includes(icon) && (
          <div className={styles.TextInputIconWrapper}>
            <FontAwesomeIcon
              icon={[VALID_ICONS[icon].prefix, VALID_ICONS[icon].name]}
              className={styles.TextInputIcon}
            />
          </div>
        )}
      </div>
      {getError(currentError, touched)}
    </>
  )
}

PrivateTextInput.PUBLIC_PROPS = {
  type: PropTypes.string,
  'data-tid': PropTypes.string.isRequired,
  disabled: PropTypes.bool,
  name: PropTypes.string.isRequired,
  allCaps: PropTypes.bool,
  maxLength: PropTypes.number,
  /** text transform capitalize label */
  capitalize: PropTypes.bool,
  initialValue: PropTypes.string,
  labelCopy: PropTypes.string,
  labelColor: PropTypes.string,
  labelWeight: PropTypes.string,
  labelClasses: PropTypes.string,
  validator: PropTypes.func,
  placeholder: PropTypes.string,
  onBlur: PropTypes.func,
  onFocus: PropTypes.func,
  restrictIllegal: PropTypes.bool,
  autoComplete: PropTypes.string,
  classOverrides: PropTypes.string,
  /** iconPrefix and iconName work together to render icon in input. Please refer to https://fontawesome.com/v5/docs/apis/javascript/import-icons for more information about iconPrefix. Please refer to `fa.js` and https://fontawesome.com for more info about icon's name. Currently allowed icons are defined by VALID_ICONS at src/helpers/constants.js */
  icon: PropTypes.oneOf(Object.keys(VALID_ICONS)),
}

PrivateTextInput.propTypes = {
  ...PrivateTextInput.PUBLIC_PROPS,
  /** text transform capitalize label */
  capitalize: PropTypes.bool,
  /** iconPrefix and iconName work together to render icon in input. Please refer to https://fontawesome.com/v5/docs/apis/javascript/import-icons for more information about iconPrefix. Please refer to `fa.js` and https://fontawesome.com for more info about icon's name. Currently allowed icons are defined by VALID_ICONS at src/helpers/constants.js */
  icon: PropTypes.oneOf(Object.keys(VALID_ICONS)),
}

PrivateTextInput.defaultProps = {
  type: 'text',
  placeholder: '',
  restrictIllegal: true,
}

function TextInputFactory(privateProps) {
  const PublicTextInputComponent = (downstreamProps) => {
    return <PrivateTextInput {...downstreamProps} {...privateProps} />
  }

  return PublicTextInputComponent
}

export const TextInput = TextInputFactory()
