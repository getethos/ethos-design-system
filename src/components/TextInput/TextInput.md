```jsx
import validateMinMaxFactory from '../../validators/validateMinMax'
import validateExists from '../../validators/validateExists'
// formChangeHandler gets wired up automatically if using <Form /> component
const formChangeHandlerStub = () => {}

;<TextInput
  name="example"
  labelCopy="Validation happens after first blur ('touched')—Value's length % 2"
  data-tid="the-text-input"
  formChangeHandler={formChangeHandlerStub}
  validator={(x) => {
    const truthyErr = validateExists(x)
    if (!!truthyErr) return truthyErr
    const minMaxErr = validateMinMaxFactory(5, 20)(x)
    if (!!minMaxErr) return minMaxErr
    return x.length % 2 ? 'Text does not have an even number of characters' : ''
  }}
/>
```

This one sets an `intialValue` which results in the field being considered as
already `touched`. This means you do not have to `blur` for field hint error
messages to appear.

```jsx
import validateMinMaxFactory from '../../validators/validateMinMax'
import validateExists from '../../validators/validateExists'
// formChangeHandler gets wired up automatically if using <Form /> component
const formChangeHandlerStub = () => {}

;<TextInput
  initialValue="hi"
  name="example"
  labelCopy="Validation happens after first blur ('touched')—Value's length % 2"
  data-tid="the-text-input"
  formChangeHandler={formChangeHandlerStub}
  validator={(x) => {
    const truthyErr = validateExists(x)
    if (!!truthyErr) return truthyErr
    const minMaxErr = validateMinMaxFactory(5, 20)(x)
    if (!!minMaxErr) return minMaxErr
    return x.length % 2 ? 'Text does not have an even number of characters' : ''
  }}
/>
```
