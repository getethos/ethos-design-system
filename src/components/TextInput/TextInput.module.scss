/* TODO -- these are currently come in via globals but we'll need
   them here once we continue to strip 'em out of design-system.css
@import "../Colors.scss";
@import "../Type/Type.scss";
@import '../Spacer/Spacer.scss';
*/

.TextInput {
  /* Same as TypeBase.scss Body. Should we use a mixin? TODO use --base-font-size */
  $font-size: 19;
  $line-height: 24;
  font-size: $font-size * 1px;
  line-height: $line-height / $font-size;
  width: 100%;
  font-family: var(--Theinhardt-font-stack);
  box-sizing: border-box;
  background-color: var(--White);
  border-style: solid;
  border-width: 1px;
  border-color: var(--GrayStrokeAndDisabled--translucent);
  padding: 17px var(--Space-16) 13px;
  color: var(--GrayPrimary--opaque);
  border-radius: 0;
  transition: border-color 250ms ease;
  outline: 0;
  margin: 0;
}

.TextInput:hover {
  background-color: var(--GrayLightHover--translucent);
}

.TextInput:focus {
  border-color: var(--GrayPrimary--opaque);
  background-color: var(--White);
}

.TextInput::placeholder {
  /* It's $gray-64 / #a3a3a3 in ethos, but going to try to limit # of grays */
  color: var(--GrayStrokeAndDisabled--opaque);
}

.TextInput:disabled {
  color: var(--White);
  background-color: var(--GrayStrokeAndDisabled--translucent);
  border-color: transparent;
}
