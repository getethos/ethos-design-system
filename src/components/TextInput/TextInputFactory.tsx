import * as React from 'react'
import { PrivateTextInput } from './PrivateTextInput'
export function TextInputFactory(privateProps) {
  const PublicTextInputComponent = (downstreamProps) => {
    return <PrivateTextInput {...downstreamProps} {...privateProps} />
  }
  return PublicTextInputComponent
}
