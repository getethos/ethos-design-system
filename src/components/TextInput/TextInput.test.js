import React from 'react'
import { TextInput } from './TextInput'
import renderer from 'react-test-renderer'

describe('TextInput', () => {
  it('default rendering', () => {
    const tree = renderer
      .create(
        <TextInput
          name="nombre"
          allCaps={true}
          labelCopy="The label"
          data-tid="the-text-input"
          validator={(x) =>
            x.length % 2
              ? 'Text does not have an even number of characters'
              : ''
          }
          autoComplete="name"
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
