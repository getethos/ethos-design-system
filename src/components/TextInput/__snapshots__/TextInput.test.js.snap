// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput default rendering 1`] = `
Array [
  <label
    className="Caption Theinhardt Medium500 GrayPrimary AllCaps"
    htmlFor="nombre"
  >
    The label
  </label>,
  <div
    aria-hidden={true}
    className="Spacer"
    style={
      Object {
        "height": 4,
      }
    }
  />,
  <div
    className="TextInputWrapper"
  >
    <input
      aria-label="nombre"
      autoComplete="name"
      className="TextInputCommon TextInputStylable"
      data-tid="the-text-input"
      name="nombre"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder=""
      type="text"
      value=""
    />
  </div>,
  "",
]
`;
